test/bench/synthetic_benchmarks.cpp

```diff
@@ -0,0 +1,241 @@
+        return InstructionKind::nullop;
+    else if (trait.stack_height_required == 1 && trait.stack_height_change == 0)
+        return InstructionKind::unop;
+    else if (trait.stack_height_required == 2 && trait.stack_height_change == -1)
+        return InstructionKind::binop;
+    else
+        return InstructionKind::unknown;
+}
+
+struct CodeParams
+{
+    evmc_opcode opcode;
+    Mode mode;
+};
+
+/// The less-than comparison operator. Needed for std::map.
+[[maybe_unused]] inline constexpr bool operator<(const CodeParams& a, const CodeParams& b) noexcept
+{
+    return std::tuple(a.opcode, a.mode) < std::tuple(b.opcode, b.mode);
+}
+
+std::string to_string(const CodeParams& params)
+{
+    return std::string{instr::traits[params.opcode].name} + '/' +
+           static_cast<char>(get_instruction_kind(params.opcode)) +
+           std::to_string(static_cast<int>(params.mode));
+}
+
+/// Generates the EVM benchmark loop inner code for the given opcode and "mode".
+bytecode generate_loop_inner_code(CodeParams params)
+{
+    const auto [opcode, mode] = params;
+    const auto kind = get_instruction_kind(opcode);
+    switch (mode)
+    {
+    case Mode::min_stack:
+        switch (kind)
+        {
+        case InstructionKind::push:
+            // PUSH1 POP PUSH1 POP ...
+            return stack_limit * (push(opcode, {}) + OP_POP);
+        case InstructionKind::swap:
+        {
+            const auto n = opcode - OP_SWAP1 + 1;
+            // DUP1 SWAP1 SWAP1 ... POP
+            return n * OP_DUP1 + stack_limit * 2 * bytecode{opcode} + n * OP_POP;
+        }
+        case InstructionKind::dup:
+        {
+            const auto n = opcode - OP_DUP1;
+            // DUP1 DUP1 POP DUP1 POP ... POP
+            return n * OP_DUP1 + (stack_limit - n) * (bytecode{opcode} + OP_POP) + n * OP_POP;
+        }
+        case InstructionKind::nullop:
+            // CALLER POP CALLER POP ...
+            return stack_limit * (bytecode{opcode} + OP_POP);
+        case InstructionKind::nop:
+            // JUMPDEST JUMPDEST ...
+            return stack_limit * 2 * bytecode{opcode};
+        case InstructionKind::unop:
+            // DUP1 NOT NOT ... POP
+            return OP_DUP1 + stack_limit * 2 * bytecode{opcode} + OP_POP;
+        case InstructionKind::binop:
+            // DUP1 DUP1 ADD DUP1 ADD DUP1 ADD ... POP
+            return OP_DUP1 + (stack_limit - 1) * (OP_DUP1 + bytecode{opcode}) + OP_POP;
+        default:
+            break;
+        }
+        break;
+    case Mode::full_stack:
+        switch (kind)
+        {
+        case InstructionKind::push:
+            // PUSH1 PUSH1 PUSH1 ... POP POP POP ...
+            return stack_limit * push(opcode, {}) + stack_limit * OP_POP;
+        case InstructionKind::dup:
+        {
+            const auto n = opcode - OP_DUP1;
+            // DUP1 DUP1 POP DUP1 POP ... POP
+            return n * OP_DUP1 + (stack_limit - n) * bytecode{opcode} + stack_limit * OP_POP;
```

Comment:
Perhaps splitting all these tests into multiple lines could show the grouping better, but then clang-tidy will complain:
```suggestion
            return n * OP_DUP1 +
              (stack_limit - n) * bytecode{opcode} +
              stack_limit * OP_POP;
```